pkg/util/updater/updater.go

```diff
@@ -0,0 +1,367 @@
+
+	u.log.Infof("Downloaded binary file to %q", path)
+
+	valid, err := isChecksumValid(path, checksum)
+	if err != nil {
+		return "", fmt.Errorf("failed to check file %q sum: %w", path, err)
+	}
+
+	if !valid {
+		return "", fmt.Errorf("checksum is not valid")
+	}
+
+	return path, nil
+}
+
+func (u *Updater) start(path string) error {
+	currentBinaryPath := u.restartCtx.CmdPath()
+
+	toBeRemoved, err := u.updateBinary(path, currentBinaryPath)
+	if err != nil {
+		return fmt.Errorf("failed to update binary: %w", err)
+	}
+
+	u.log.Infof("Need to remove file in %q", toBeRemoved)
+
+	defer func() {
+		if err == nil {
+			go func() {
+				time.Sleep(exitDelay)
+
+				u.log.Infof("Removing file in %q", toBeRemoved)
+
+				if err := os.Remove(toBeRemoved); err != nil {
+					u.log.Errorf("Failed to remove file %q: %v", toBeRemoved, err)
+				}
+
+				u.log.Infof("Exiting")
+				os.Exit(0)
+			}()
+		}
+	}()
+
+	u.log.Infof("Starting new file instance")
+
+	if err := u.restartCtx.Start(); err != nil {
+		u.log.Errorf("Failed to restart visor: %v", err)
+
+		// Restore old binary file
+		if err := os.Remove(currentBinaryPath); err != nil {
+			u.log.Errorf("Failed to remove file %q: %v", currentBinaryPath, err)
+		}
+
+		if err := os.Rename(toBeRemoved, currentBinaryPath); err != nil {
+			u.log.Errorf("Failed to rename file %q to %q: %v", toBeRemoved, currentBinaryPath, err)
+		}
+
+		return fmt.Errorf("failed to restart visor: %w", err)
+	}
+
+	return nil
+}
+
+func (u *Updater) updateBinary(downloadPath, currentPath string) (toBeRemoved string, err error) {
+	oldPath := currentPath + ".old"
+
+	if _, err := os.Stat(oldPath); err == nil {
+		if err := os.Remove(oldPath); err != nil {
+			return "", err
+		}
+	}
+
+	if err := os.Rename(currentPath, oldPath); err != nil {
+		return "", err
+	}
+
+	if err := os.Rename(downloadPath, currentPath); err != nil {
+		// Try to revert previous os.Rename
+		if err := os.Rename(oldPath, currentPath); err != nil {
+			u.log.Errorf("Failed to rename file %q to %q: %v", oldPath, currentPath, err)
+		}
+
+		return "", err
+	}
+
+	return oldPath, nil
+}
+
+func isChecksumValid(filename, wantSum string) (bool, error) {
+	f, err := os.Open(filepath.Clean(filename))
+	if err != nil {
+		return false, err
+	}
+
+	hasher := sha256.New()
+	if _, err := io.Copy(hasher, f); err != nil {
+		return false, err
+	}
+
+	if err := f.Close(); err != nil {
+		return false, err
+	}
+
+	gotSum := hex.EncodeToString(hasher.Sum(nil))
+
+	return gotSum == wantSum, nil
+}
+
+func getChecksum(checksums, filename string) (string, error) {
+	idx := strings.Index(checksums, filename)
+	if idx == -1 {
+		return "", ErrNoChecksumFound
+	}
+
+	last := idx - 1 // space separator
+	first := last - checkSumLength
+
+	if first < 0 {
+		return "", ErrMalformedChecksumFile
+	}
+
+	return checksums[first:last], nil
+}
+
+func downloadChecksums(url string) (checksums string, err error) {
+	resp, err := http.Get(url) // nolint:gosec
+	if resp != nil {
```

Comment:
I've seen some advice that recommends checking the response body even if `err != nil` because it may happen on a redirection failure condition. E.g.: http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body

However, I've checked `(*Client) Do` in `net/http/client.go` and found this:
```
// On error, any Response can be ignored. A non-nil Response with a
// non-nil error only occurs when CheckRedirect fails, and even then
// the returned Response.Body is already closed.
```

So, actually, we can rewrite it as you said if you think it will be more readable. There may be similar code snippets written by me throughout the project, so in this case, we may also need to change them too.

Reply:
I think it would be great to do so, just to make things a bit cleaner